app/lib/shared/task_scheduler.dart

```diff
@@ -44,68 +40,78 @@ class TaskScheduler {
   final TaskRunner taskRunner;
   final List<TaskSource> sources;
   final LastNTracker<String> _statusTracker = new LastNTracker();
-  final LastNTracker<num> _allLatencyTracker = new LastNTracker();
-  final LastNTracker<num> _workLatencyTracker = new LastNTracker();
-  int _pendingCount = 0;
+  final LastNTracker<num> _latencyTracker = new LastNTracker();
+  List<Queue<Task>> _queues;
 
-  TaskScheduler(this.taskRunner, this.sources);
+  TaskScheduler(this.taskRunner, this.sources) {
+    _queues = new List<Queue<Task>>.generate(
+        sources.length, (i) => new Queue<Task>());
+  }
 
   Future run() async {
     Future runTask(Task task) async {
       final Stopwatch sw = new Stopwatch()..start();
       try {
-        TaskTargetStatus status;
         if (redirectPackagePages.containsKey(task.package)) {
-          status = new TaskTargetStatus.skip('Package is redirected from pub.');
-        }
-        status ??= await taskRunner.checkTargetStatus(task);
-        if (status.shouldSkip) {
-          _logger.info('Skipping scheduled task: $task: ${status.reason}.');
-          _statusTracker.add('skip');
-          _allLatencyTracker.add(sw.elapsedMilliseconds);
           return;
         }
-        final bool raceDetected = await taskRunner.runTask(task);
-        _statusTracker.add(raceDetected ? 'race' : 'normal');
+        await taskRunner.runTask(task);
+        _statusTracker.add('normal');
       } catch (e, st) {
         _logger.severe('Error processing task: $task', e, st);
         _statusTracker.add('error');
       }
-      _workLatencyTracker.add(sw.elapsedMilliseconds);
-      _allLatencyTracker.add(sw.elapsedMilliseconds);
+      _latencyTracker.add(sw.elapsedMilliseconds);
     }
 
-    final Pool pool = new Pool(1);
-    final futures = sources
-        .map((TaskSource ts) {
-          final stream = ts.startStreaming();
-          return stream.listen((task) {
-            _pendingCount++;
-            final f = pool.withResource(() => runTask(task));
-            f.whenComplete(() {
-              _pendingCount--;
-            });
-          });
-        })
-        .map((subscription) => subscription.asFuture())
-        .toList();
-    await Future.wait(futures);
+    int liveSubscriptions = sources.length;
+    for (int i = 0; i < sources.length; i++) {
+      sources[i].startStreaming().listen(
+        (task) {
+          _queues[i].add(task);
+        },
+        onDone: () {
+          liveSubscriptions--;
+        },
+      );
+    }
+
+    for (;;) {
+      final task = _pickFirstTask();
+
+      if (task == null) {
+        if (liveSubscriptions == 0) {
+          break;
+        } else {
+          await new Future.delayed(const Duration(seconds: 5));
```

Comment:
I like to have a 'bool running' (or a Future / Completer depending on the design) and then set it to true when the queue is being processed. Once the queue is done, I return. When an element is added to the queue, I respawn the worker future if it's not running.

Alternatively, is there a way to merge the streams into a single stream? Then this could be a big single await for. It would also be fair and handle when all the subscriptions are dead.

Reply:
I've updated the for loop, and I could rewrite this with some Completer-based notification from the stream, but I think the later would not add more clarity on what's going on.

Reply:
Well this is a polling loop, those are generally bad (power inefficient and bad reaction time), but okay